File: lessons/dsa/prefix-sums/examples/range_sums.py
"""Range totals, computed directly and through prefix sums.

The prefix array carries a leading zero, so `prefix[i]` is the total of the
first i values and a range needs no special case when it starts at position 0.
"""


def build_prefix(values: list[int]) -> list[int]:
    """One pass. Entry i holds the total of the first i values."""
    prefix = [0]
    for value in values:
        prefix.append(prefix[-1] + value)
    return prefix


def range_total_by_scan(values: list[int], start: int, end: int) -> int:
    """Add the range every time it is asked for."""
    total = 0
    for index in range(start, end + 1):
        total += values[index]
    return total


def range_total_by_prefix(prefix: list[int], start: int, end: int) -> int:
    """Two reads and a subtraction, whatever the range covers."""
    return prefix[end + 1] - prefix[start]

File: lessons/dsa/prefix-sums/examples/range-sums.ts
/**
 * Range totals, computed directly and through prefix sums.
 *
 * The prefix array carries a leading zero, so `prefix[i]` is the total of the
 * first i values and a range needs no special case when it starts at 0.
 */

/** One pass. Entry i holds the total of the first i values. */
export function buildPrefix(values: readonly number[]): number[] {
  const prefix = [0];
  for (const value of values) {
    prefix.push((prefix.at(-1) ?? 0) + value);
  }
  return prefix;
}

/** Add the range every time it is asked for. */
export function rangeTotalByScan(
  values: readonly number[],
  start: number,
  end: number,
): number {
  let total = 0;
  for (let index = start; index <= end; index += 1) {
    total += values[index] ?? 0;
  }
  return total;
}

/** Two reads and a subtraction, whatever the range covers. */
export function rangeTotalByPrefix(
  prefix: readonly number[],
  start: number,
  end: number,
): number {
  return (prefix[end + 1] ?? 0) - (prefix[start] ?? 0);
}
